system/lib/pthread/library_pthread.c

```diff
@@ -214,10 +231,52 @@ static void _do_call(em_queued_call *q)
 }
 
 #define CALL_QUEUE_SIZE 128
-static em_queued_call **call_queue = 0;
-static volatile int call_queue_head = 0; // Shared data synchronized by call_queue_lock.
-static volatile int call_queue_tail = 0;
+
+typedef struct CallQueue
+{
+	void *target_thread;
+	em_queued_call **call_queue;
+	volatile int call_queue_head; // Shared data synchronized by call_queue_lock.
```

Comment:
Let's not use volatile for thread-shared data. There is a possibility that this will end up not even being compiled to atomic in upstream LLVM (see https://github.com/WebAssembly/tool-conventions/issues/60 and feel free to express any opinions you have about that).

Instead we can use `_Atomic int`

Reply:
`volatile` here is used in the pure C meaning "all reads and writes to the shared data must be strictly observed". The keyword is not used as synchronization primitive. I do not care about actual codegen being atomic here. It is also used documentationally to say "this is the shared data access point".

In Firefox we had issues in asm.js multithreading with

```c
int data = 0;
volatile int *spinlock = 0;
void foo()
{
  printf("data: %d\n", data);
  while(emscripten_atomic_exchange_u32(&lock, 1))
    ;
  printf("data: %d\n", data);
}
```

transformed to the following:

```c
int data = 0;
volatile int *spinlock = 0;
void foo()
{
  int d = data;
  printf("data: %d\n", d);
  while(emscripten_atomic_exchange_u32(&lock, 1))
    ;
  printf("data: %d\n", d);
}
```

I don't recall if this occurred in C compiler, with fastcomp not realizing that `emscripten_atomic_exchange_u32()` should be a full barrier, or with Firefox JS engine doing such transformation at JS runtime (I think at some point we had both happening).

In wasm multithreading things are considerably better, and `volatile` here should mean practically nothing in terms of synchronization, so should not hurt to have them here.

The actual synchronization is done with pthread mutex locks (for now, which I think could be optimized away), and/or with actual atomic operations.